File: go/structs/linkedlist/linkedlist.go
package linkedlist

import (
	"encoding/json"
	"fmt"
)

// Node represents a linked list node
type Node struct {
	Data string
	Next *Node
}

// LinkedList represents a singly linked list
type LinkedList struct {
	head *Node
	size int
}

// New creates a new LinkedList
func New() *LinkedList {
	return &LinkedList{
		head: nil,
		size: 0,
	}
}

// PushFront adds an element to the front
func (l *LinkedList) PushFront(value string) {
	newNode := &Node{
		Data: value,
		Next: l.head,
	}
	l.head = newNode
	l.size++
}

// PushBack adds an element to the back
func (l *LinkedList) PushBack(value string) {
	newNode := &Node{
		Data: value,
		Next: nil,
	}

	if l.head == nil {
		l.head = newNode
	} else {
		current := l.head
		for current.Next != nil {
			current = current.Next
		}
		current.Next = newNode
	}
	l.size++
}

// Insert adds an element at specified index
func (l *LinkedList) Insert(index int, value string) error {
	if index < 0 || index > l.size {
		return fmt.Errorf("index out of range: %d", index)
	}

	if index == 0 {
		l.PushFront(value)
		return nil
	}

	newNode := &Node{Data: value}
	current := l.head
	for i := 0; i < index-1; i++ {
		current = current.Next
	}
	newNode.Next = current.Next
	current.Next = newNode
	l.size++
	return nil
}

// Get returns element at index
func (l *LinkedList) Get(index int) (string, error) {
	if index < 0 || index >= l.size {
		return "", fmt.Errorf("index out of range: %d", index)
	}

	current := l.head
	for i := 0; i < index; i++ {
		current = current.Next
	}
	return current.Data, nil
}

// Set updates element at index
func (l *LinkedList) Set(index int, value string) error {
	if index < 0 || index >= l.size {
		return fmt.Errorf("index out of range: %d", index)
	}

	current := l.head
	for i := 0; i < index; i++ {
		current = current.Next
	}
	current.Data = value
	return nil
}

// PopFront removes the first element
func (l *LinkedList) PopFront() error {
	if l.head == nil {
		return fmt.Errorf("list is empty")
	}

	l.head = l.head.Next
	l.size--
	return nil
}

// Remove deletes element at index
func (l *LinkedList) Remove(index int) error {
	if index < 0 || index >= l.size {
		return fmt.Errorf("index out of range: %d", index)
	}

	if index == 0 {
		return l.PopFront()
	}

	current := l.head
	for i := 0; i < index-1; i++ {
		current = current.Next
	}
	current.Next = current.Next.Next
	l.size--
	return nil
}

// RemoveByValue removes first occurrence of value
func (l *LinkedList) RemoveByValue(value string) bool {
	if l.head == nil {
		return false
	}

	if l.head.Data == value {
		l.PopFront()
		return true
	}

	current := l.head
	for current.Next != nil {
		if current.Next.Data == value {
			current.Next = current.Next.Next
			l.size--
			return true
		}
		current = current.Next
	}
	return false
}

// Size returns the number of elements
func (l *LinkedList) Size() int {
	return l.size
}

// IsEmpty checks if list is empty
func (l *LinkedList) IsEmpty() bool {
	return l.size == 0
}

// Clear removes all elements
func (l *LinkedList) Clear() {
	l.head = nil
	l.size = 0
}

// ToSlice returns all elements as slice
func (l *LinkedList) ToSlice() []string {
	result := make([]string, 0, l.size)
	current := l.head
	for current != nil {
		result = append(result, current.Data)
		current = current.Next
	}
	return result
}

// MarshalJSON implements json.Marshaler
func (l *LinkedList) MarshalJSON() ([]byte, error) {
	return json.Marshal(l.ToSlice())
}

// UnmarshalJSON implements json.Unmarshaler
func (l *LinkedList) UnmarshalJSON(data []byte) error {
	var items []string
	if err := json.Unmarshal(data, &items); err != nil {
		return err
	}

	l.Clear()
	for _, item := range items {
		l.PushBack(item)
	}
	return nil
}

File: C++/src/LinkedList.cpp
#include "../include/LinkedList.h"
#include <stdexcept>

LinkedList::LinkedList() : head(nullptr), size(0) {}

LinkedList::LinkedList(const LinkedList &other) : head(nullptr), size(0) {
  Node *current = other.head;
  while (current != nullptr) {
    pushBack(current->data);
    current = current->next;
  }
}

LinkedList::~LinkedList() { clear(); }

LinkedList &LinkedList::operator=(const LinkedList &other) {
  if (this != &other) {
    clear();
    Node *current = other.head;
    while (current != nullptr) {
      pushBack(current->data);
      current = current->next;
    }
  }
  return *this;
}

void LinkedList::pushFront(const std::string &value) {
  Node *newNode = new Node(value);
  newNode->next = head;
  head = newNode;
  size++;
}

void LinkedList::pushBack(const std::string &value) {
  Node *newNode = new Node(value);
  if (head == nullptr) {
    head = newNode;
  } else {
    Node *current = head;
    while (current->next != nullptr) {
      current = current->next;
    }
    current->next = newNode;
  }
  size++;
}

void LinkedList::insert(int index, const std::string &value) {
  if (index < 0 || index > size) {
    throw std::out_of_range("Index out of range");
  }
  if (index == 0) {
    pushFront(value);
    return;
  }

  Node *newNode = new Node(value);
  Node *current = head;
  for (int i = 0; i < index - 1; i++) {
    current = current->next;
  }
  newNode->next = current->next;
  current->next = newNode;
  size++;
}

std::string LinkedList::get(int index) const {
  if (index < 0 || index >= size) {
    throw std::out_of_range("Index out of range");
  }
  Node *current = head;
  for (int i = 0; i < index; i++) {
    current = current->next;
  }
  return current->data;
}

void LinkedList::set(int index, const std::string &value) {
  if (index < 0 || index >= size) {
    throw std::out_of_range("Index out of range");
  }
  Node *current = head;
  for (int i = 0; i < index; i++) {
    current = current->next;
  }
  current->data = value;
}

void LinkedList::popFront() {
  if (head == nullptr) {
    throw std::runtime_error("List is empty");
  }
  Node *temp = head;
  head = head->next;
  delete temp;
  size--;
}

void LinkedList::popBack() {
  if (head == nullptr) {
    throw std::runtime_error("List is empty");
  }
  if (head->next == nullptr) {
    delete head;
    head = nullptr;
    size--;
    return;
  }

  Node *current = head;
  while (current->next->next != nullptr) {
    current = current->next;
  }
  delete current->next;
  current->next = nullptr;
  size--;
}

void LinkedList::remove(int index) {
  if (index < 0 || index >= size) {
    throw std::out_of_range("Index out of range");
  }
  if (index == 0) {
    popFront();
    return;
  }

  Node *current = head;
  for (int i = 0; i < index - 1; i++) {
    current = current->next;
  }
  Node *temp = current->next;
  current->next = temp->next;
  delete temp;
  size--;
}

bool LinkedList::removeByValue(const std::string &value) {
  if (head == nullptr)
    return false;

  if (head->data == value) {
    popFront();
    return true;
  }

  Node *current = head;
  while (current->next != nullptr) {
    if (current->next->data == value) {
      Node *temp = current->next;
      current->next = temp->next;
      delete temp;
      size--;
      return true;
    }
    current = current->next;
  }
  return false;
}

void LinkedList::clear() {
  while (head != nullptr) {
    Node *temp = head;
    head = head->next;
    delete temp;
  }
  size = 0;
}

void LinkedList::serialize(std::ofstream &out) const {
  out.write(reinterpret_cast<const char *>(&size), sizeof(size));
  Node *current = head;
  while (current != nullptr) {
    int len = current->data.length();
    out.write(reinterpret_cast<const char *>(&len), sizeof(len));
    out.write(current->data.c_str(), len);
    current = current->next;
  }
}

void LinkedList::deserialize(std::ifstream &in) {
  clear();
  int newSize;
  in.read(reinterpret_cast<char *>(&newSize), sizeof(newSize));

  for (int i = 0; i < newSize; i++) {
    int len;
    in.read(reinterpret_cast<char *>(&len), sizeof(len));
    char *buffer = new char[len + 1];
    in.read(buffer, len);
    buffer[len] = '\0';
    pushBack(std::string(buffer));
    delete[] buffer;
  }
}
